### PR TITLE
[Merged by Bors] - feat: Rank-nullity theorem for commutative domains

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/Submodule/Localization.lean [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+universe u u' v v'
+
+variable {R : Type u} (S : Type u') {M : Type v} {N : Type v'}
+variable [CommRing R] [CommRing S] [AddCommGroup M] [AddCommGroup N] [Module R M]
+variable [Module R M] [Module R N] [Algebra R S] [Module S N] [IsScalarTower R S N]
+variable (p : Submonoid R) [IsLocalization p S] (f : M →ₗ[R] N) [IsLocalizedModule p f]
+variable (hp : p ≤ R⁰)
+
+variable {S} in
+lemma IsLocalizedModule.mk'_smul_mk' {x : R} {m : M} {s t : p} :
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Module/Submodule/Localization.lean:36]: Is there a reason all variables are implicit here?

File: Mathlib/Algebra/Module/Submodule/Localization.lean [hunk 2]
```diff
@@ -0,0 +1,121 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.Algebra.Module.LocalizedModule
+
+/-!
+# Localization of Submodules
+
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Module/Submodule/Localization.lean:10]: Can you add here a brief description of the file?

File: Mathlib/LinearAlgebra/Dimension/Constructions.lean [hunk 3]
```diff
@@ -449,9 +428,93 @@ theorem Submodule.finrank_le_finrank_of_le {s t : Submodule R M} [Module.Finite
...
+  finrank_le_of_rank_le (by simpa using rank_span_le (R := R) s)
+#align finrank_span_le_card finrank_span_le_card
+
+theorem finrank_span_finset_le_card (s : Finset M) : (s : Set M).finrank R ≤ s.card :=
+  calc
+    (s : Set M).finrank R ≤ (s : Set M).toFinset.card := finrank_span_le_card (M := M) s
+    _ = s.card := by simp
+#align finrank_span_finset_le_card finrank_span_finset_le_card
+
+theorem finrank_range_le_card {ι : Type*} [Fintype ι] {b : ι → M} :
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Dimension/Constructions.lean:472]: Here I think `b` should be explicit.

File: Mathlib/LinearAlgebra/Dimension/Constructions.lean [hunk 4]
```diff
@@ -449,9 +428,93 @@ theorem Submodule.finrank_le_finrank_of_le {s t : Submodule R M} [Module.Finite
...
+    (hb : LinearIndependent R b) :
+    finrank R (span R (Set.range b)) = Fintype.card ι :=
+  finrank_eq_of_rank_eq
+    (by
+      have : Module.rank R (span R (Set.range b)) = #(Set.range b) := rank_span hb
+      rwa [← lift_inj, mk_range_eq_of_injective hb.injective, Cardinal.mk_fintype, lift_natCast,
+        lift_eq_nat_iff] at this)
+#align finrank_span_eq_card finrank_span_eq_card
+
+theorem finrank_span_set_eq_card (s : Set M) [Fintype s] (hs : LinearIndependent R ((↑) : s → M)) :
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Dimension/Constructions.lean:490]:
```suggestion
theorem finrank_span_set_eq_card {s : Set M} [Fintype s] (hs : LinearIndependent R ((↑) : s → M)) :
```
Feel free to keep it explicit if you think it is better.

Same below.

File: Mathlib/LinearAlgebra/Dimension/Localization.lean [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    simp_rw [smul_zero, Finset.smul_sum, ← mul_smul, Submonoid.smul_def, ← mul_smul, mul_comm] at hc
+    simp only [hsec, zero_mul, map_eq_zero_iff (algebraMap R S) (IsLocalization.injective S hp)]
+    apply hp (c * u i).prop
+    exact hs t _ hc _ hit
+
+lemma IsLocalizedModule.rank_eq {N : Type v} [AddCommGroup N]
+    [Module R N] [Module S N] [IsScalarTower R S N] (f : M →ₗ[R] N) [IsLocalizedModule p f] :
+    Module.rank S N = Module.rank R M := by simpa using IsLocalizedModule.lift_rank_eq S p f hp
+
+theorem rank_quotient_add_rank_of_isDomain [IsDomain R] (M' : Submodule R M) :
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Dimension/Localization.lean:77]: Can you add a docstring here? Thanks!

**Review by riccardobrasca — APPROVED** [review]
Can you please add a description of the PR?

Thanks!

**Review by riccardobrasca — APPROVED** [review]
Can you add a reference to the version for domains to the docstring of [rank_range_add_rank_ker](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Dimension/DivisionRing.html#rank_range_add_rank_ker)?

Also I think this (or the version over a field) should be added to https://leanprover-community.github.io/undergrad.html and https://leanprover-community.github.io/mathlib-overview.html

Thanks!!

bors d+

**Comment by erdOne** [comment]
bors merge

**Comment by erdOne** [comment]
I think I'll add https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/finite_dimensional.20and.20finite/near/411069786 in a follow up PR and the docstring of both lemmas and undergrad.yml should all point there.

**Comment by erdOne** [comment]
bors merge

**Pull request closed without merging**